ratpack-core/src/main/java/ratpack/exec/util/Batch.java

```diff
@@ -0,0 +1,365 @@
+ *         .parallel()
+ *         .publisher()
+ *         .reduce(0, (r, i) -> r + i);
+ *
+ *     }).getValueOrThrow();
+ *
+ *     assertEquals(sum, 4950);
+ *   }
+ * }
+ * }</pre>
+ *
+ * @param <T> the type of item in the batch
+ * @since 1.4
+ * @see #yieldAll()
+ * @see #publisher()
+ */
+public final class Batch<T> {
+
+  private final Iterable<? extends Promise<T>> jobs;
+  private final Action<? super Execution> execInit;
+
+  private Batch(Iterable<? extends Promise<T>> jobs, @Nullable Action<? super Execution> execInit) {
+    this.jobs = jobs;
+    this.execInit = execInit;
+  }
+
+  /**
+   * Creates a new batch of the given promises.
+   * <p>
+   * The given iterable will not be consumed until processing begins.
+   *
+   * @param jobs the batch jobs
+   * @param <T> the type of item produced by each job
+   * @return a new batch
+   */
+  public static <T> Batch<T> of(Iterable<? extends Promise<T>> jobs) {
+    return new Batch<>(jobs, null);
+  }
+
+  /**
+   * Creates a new batch of the given promises.
+   *
+   * @param jobs the batch jobs
+   * @param <T> the type of item produced by each job
+   * @return a new batch
+   */
+  @SafeVarargs
+  @SuppressWarnings("varargs")
+  public static <T> Batch<T> of(Promise<T>... jobs) {
+    return new Batch<>(Arrays.asList(jobs), null);
+  }
+
+  /**
+   * Specifies that the batch should be processed in parallel.
+   *
+   * @return a parallel batch
+   * @see #parallel(Action)
+   */
+  public Batch<T> parallel() {
+    return parallel(Action.noop());
+  }
+
+  /**
+   * Specifies that the batch should be processed in parallel, and the given action should initialize each execution.
+   * <p>
+   * Each promise will be subscribed to in a {@link Execution#fork() forked execution}.
+   * The given action will be called with each execution before processing the promise.
+   * This can be used to seed the execution registry.
+   *
+   * @param executionInitializer the execution initializer
+   * @return a parallel batch
+   */
+  public Batch<T> parallel(Action<? super Execution> executionInitializer) {
+    return new Batch<>(jobs, executionInitializer);
+  }
+
+  /**
+   * Processes the batch jobs, return {@link Result} objects for each.
+   * <p>
+   * Every job will be processed, regardless of any failure.
+   * <p>
+   * The order of the returned list corresponds to the original job order.
+   * That is, it is guaranteed that the 2nd item in the list was the 2nd job specified.
+   * It does not reflect the order in which jobs completed.
+   *
+   * @return a promise for the result of each job
+   */
+  public Promise<List<? extends ExecResult<T>>> yieldAll() {
+    List<Promise<T>> jobs = Lists.newArrayList(this.jobs);
+    if (jobs.isEmpty()) {
+      return Promise.value(Collections.emptyList());
+    }
+
+    return execInit == null ? serialYieldAll(jobs) : parallelYieldAll(jobs, execInit);
+  }
+
+  /**
+   * Processes the batch jobs, return {@link Result} objects for each.
```

Comment:
This method is actually returning Promise for a list of promised values, not Result objects.